File: packages/php/tests/ConformanceTest.php
<?php

declare(strict_types=1);

namespace OrbitId\Tests;

use OrbitId\OrbitError;
use OrbitId\OrbitGenerator;
use OrbitId\OrbitId;
use PHPUnit\Framework\TestCase;

final class ConformanceTest extends TestCase
{
    public function testEncodeDecodeFixtures(): void
    {
        foreach ($this->fixture('encode-decode.v1.json')['cases'] as $case) {
            $fields = [
                'timestamp' => $case['timestamp'],
                'type' => $case['type'],
                'node' => $case['node'],
                'sequence' => $case['sequence'],
            ];
            $id = OrbitId::encode($fields);

            self::assertSame($case['idDecimal'], $id, $case['id']);
            self::assertSame(strtolower($case['idHex']), OrbitId::toHexString($id), $case['id']);
            self::assertSame($fields, OrbitId::decode($id), $case['id']);
            self::assertSame($fields, OrbitId::parse($case['idDecimal']), $case['id']);
            self::assertSame($case['timestamp'], OrbitId::getTimestamp($id), $case['id']);
            self::assertSame($case['type'], OrbitId::getType($id), $case['id']);
            self::assertSame($case['node'], OrbitId::getNode($id), $case['id']);
            self::assertSame($case['sequence'], OrbitId::getSequence($id), $case['id']);
            self::assertTrue(OrbitId::isValid($id), $case['id']);
        }
    }

    public function testRejectFixtures(): void
    {
        foreach ($this->fixture('decode-reject.v1.json')['cases'] as $case) {
            try {
                OrbitId::fromDecimalString($case['input']);
                self::fail("Expected INVALID_DECIMAL for {$case['id']}");
            } catch (OrbitError $error) {
                self::assertSame(OrbitError::INVALID_DECIMAL, $error->orbitCode, $case['id']);
            }
            self::assertFalse(OrbitId::isValid($case['input']), $case['id']);
        }
        self::assertSame('0', OrbitId::fromDecimalString('0'));
    }

    public function testGeneratorFixtures(): void
    {
        $fixture = $this->fixture('generator.v1.json');
        foreach ($fixture['cases'] as $case) {
            $generator = new OrbitGenerator([
                'node' => $case['node'],
                'clockRollbackToleranceMs' => (int) $fixture['defaults']['clockRollbackToleranceMs'],
                'onSequenceExhausted' => 'fail',
                'clock' => static fn(): string => $case['nowTimestamp'],
            ]);
            $generator->restoreState($case['prior']['lastTimestamp'], $case['prior']['sequence']);
            $decision = $generator->decide($case['type'], $case['nowTimestamp']);
            $expected = $case['expect'];

            if ($expected['action'] === 'issue') {
                self::assertSame([
                    'action' => 'issue',
                    'timestamp' => $expected['timestamp'],
                    'sequence' => $expected['sequence'],
                ], $decision, $case['id']);
            } elseif ($expected['action'] === 'wait') {
                self::assertSame([
                    'action' => 'wait',
                    'waitUntilTimestamp' => $expected['waitUntilTimestamp'],
                ], $decision, $case['id']);
            } elseif ($expected['action'] === 'wait_or_fail') {
                self::assertSame('error', $decision['action'], $case['id']);
                self::assertSame($expected['error'], $decision['error'], $case['id']);
            } else {
                self::assertSame([
                    'action' => 'error',
                    'error' => $expected['error'],
                ], $decision, $case['id']);
            }
        }
    }

    public function testGeneratorCanChooseWaitAfterSequenceExhaustion(): void
    {
        $generator = new OrbitGenerator([
            'node' => 7,
            'onSequenceExhausted' => 'wait',
            'clock' => static fn(): int => 1000,
        ]);
        $generator->restoreState(1000, 1023);
        self::assertSame([
            'action' => 'wait_next_ms',
            'fromTimestamp' => '1000',
        ], $generator->decide(1, 1000));
    }

    public function testGenerateHelpersAndFunctions(): void
    {
        $ticks = [1000, 1000, 1001, 1001];
        $index = 0;
        $generator = new OrbitGenerator([
            'node' => 7,
            'onSequenceExhausted' => 'wait',
            'clock' => static function () use (&$ticks, &$index): int {
                $value = $ticks[min($index, count($ticks) - 1)];
                $index++;
                return $value;
            },
        ]);
        self::assertSame(0, $generator->getSequence());
        $id = $generator->generate(1);
        self::assertNotSame('', $id);
        self::assertGreaterThan('0', $generator->getLastTimestamp());

        $waitIndex = 0;
        $waitTicks = [1000, 1000, 1001, 1001];
        $waiter = new OrbitGenerator([
            'node' => 7,
            'onSequenceExhausted' => 'wait',
            'clock' => static function () use (&$waitTicks, &$waitIndex): int {
                $value = $waitTicks[min($waitIndex, count($waitTicks) - 1)];
                $waitIndex++;
                return $value;
            },
        ]);
        $waiter->restoreState(1000, 1023);
        self::assertNotSame('', $waiter->generate(1));
        self::assertSame('1001', (string) $waiter->getLastTimestamp());

        $sample = '140612821619842090';
        self::assertSame('16762354567', \OrbitId\getTimestamp($sample));
        self::assertSame(2, \OrbitId\getType($sample));
        self::assertSame(7, \OrbitId\getNode($sample));
        self::assertSame(42, \OrbitId\getSequence($sample));
        self::assertTrue(\OrbitId\isValid($sample));
        self::assertSame($sample, \OrbitId\toDecimalString($sample));
        self::assertSame($sample, \OrbitId\fromDecimalString($sample));
        self::assertSame(strtolower(OrbitId::toHexString($sample)), \OrbitId\toHexString($sample));
        self::assertSame(OrbitId::decode($sample), \OrbitId\decode($sample));
        self::assertSame(OrbitId::parse($sample), \OrbitId\parse($sample));
        self::assertSame(OrbitId::encode([
            'timestamp' => '16762354567',
            'type' => 2,
            'node' => 7,
            'sequence' => 42,
        ]), \OrbitId\encode([
            'timestamp' => '16762354567',
            'type' => 2,
            'node' => 7,
            'sequence' => 42,
        ]));
        $unix = \OrbitId\toUnixTimeMs(0);
        self::assertSame('0', \OrbitId\fromUnixTimeMs($unix));
        self::assertIsCallable(\OrbitId\systemOrbitClock());

        $lost = new OrbitGenerator([
            'node' => 1,
            'clock' => static fn(): int => 5,
            'confirmOwnership' => static fn(): bool => false,
        ]);
        $this->expectException(OrbitError::class);
        $lost->generate(1);
    }

    /** @return array<string, mixed> */
    private function fixture(string $name): array
    {
        $path = dirname(__DIR__, 3) . '/spec/conformance/' . $name;
        $fixture = json_decode((string) file_get_contents($path), true, 512, JSON_THROW_ON_ERROR);
        self::assertIsArray($fixture);
        return $fixture;
    }
}

File: packages/java/src/test/java/com/github/orbitid/ConformanceTest.java
package com.github.orbitid;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Path;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ConformanceTest {
    private static final ObjectMapper JSON = new ObjectMapper();
    private static final Path FIXTURES = Path.of("..", "..", "spec", "conformance");

    @Test
    void encodeDecodeFixtures() throws IOException {
        for (JsonNode testCase : fixture("encode-decode.v1.json").withArray("cases")) {
            long timestamp = Long.parseLong(testCase.get("timestamp").asText());
            int type = testCase.get("type").asInt();
            int node = testCase.get("node").asInt();
            int sequence = testCase.get("sequence").asInt();
            long id = OrbitId.encode(timestamp, type, node, sequence);

            assertEquals(testCase.get("idDecimal").asText(), OrbitId.toDecimalString(id));
            assertEquals(testCase.get("idHex").asText().toLowerCase(), OrbitId.toHexString(id));
            assertEquals(new OrbitFields(timestamp, type, node, sequence), OrbitId.decode(id));
            assertEquals(new OrbitFields(timestamp, type, node, sequence),
                    OrbitId.parse(testCase.get("idDecimal").asText()));
            assertEquals(id, OrbitId.fromDecimalString(testCase.get("idDecimal").asText()));
            assertTrue(OrbitId.isValid(id));
            assertTrue(OrbitId.isValid(testCase.get("idDecimal").asText()));
        }
    }

    @Test
    void rejectsNonCanonicalDecimals() throws IOException {
        for (JsonNode testCase : fixture("decode-reject.v1.json").withArray("cases")) {
            OrbitError error = assertThrows(OrbitError.class,
                    () -> OrbitId.fromDecimalString(testCase.get("input").asText()));
            assertEquals(OrbitError.INVALID_DECIMAL, error.getCode());
            assertFalse(OrbitId.isValid(testCase.get("input").asText()));
        }
        assertEquals(0L, OrbitId.fromDecimalString("0"));
        assertTrue(OrbitId.isValid("0"));
    }

    @Test
    void generatorFixtures() throws IOException {
        for (JsonNode testCase : fixture("generator.v1.json").withArray("cases")) {
            long now = Long.parseLong(testCase.get("nowTimestamp").asText());
            OrbitGenerator generator = new OrbitGenerator(GeneratorOptions.builder(testCase.get("node").asInt())
                    .clock(() -> now)
                    .clockRollbackToleranceMs(5_000)
                    .onSequenceExhausted(SequenceExhaustedMode.FAIL)
                    .build());
            JsonNode prior = testCase.get("prior");
            generator.restoreState(Long.parseLong(prior.get("lastTimestamp").asText()), prior.get("sequence").asInt());
            GenerateDecision decision = generator.decide(testCase.get("type").asInt(), now);
            JsonNode expected = testCase.get("expect");

            switch (expected.get("action").asText()) {
                case "issue":
                    assertEquals(new GenerateDecision.Issue(
                            Long.parseLong(expected.get("timestamp").asText()), expected.get("sequence").asInt()), decision);
                    break;
                case "wait":
                    assertEquals(new GenerateDecision.Wait(
                            Long.parseLong(expected.get("waitUntilTimestamp").asText())), decision);
                    break;
                case "error":
                    assertEquals(new GenerateDecision.Error(expected.get("error").asText()), decision);
                    break;
                case "wait_or_fail":
                    assertTrue(decision instanceof GenerateDecision.WaitNextMs
                            || decision.equals(new GenerateDecision.Error(expected.get("error").asText())));
                    break;
                default:
                    throw new AssertionError("unknown action");
            }
        }
    }

    @Test
    void coversGenerateGettersAndEncodeHelpers() {
        long[] ticks = {1000L, 1000L, 1001L, 1001L};
        int[] index = {0};
        OrbitGenerator generator = new OrbitGenerator(GeneratorOptions.builder(7)
                .clock(() -> ticks[Math.min(index[0]++, ticks.length - 1)])
                .onSequenceExhausted(SequenceExhaustedMode.WAIT)
                .build());
        assertEquals(7, generator.getNode());
        assertEquals(0L, generator.getLastTimestamp());
        assertEquals(0, generator.getSequence());

        long id = generator.generate(1);
        assertTrue(id != 0L);
        assertTrue(generator.getLastTimestamp() > 0L);

        int[] waitIndex = {0};
        long[] waitTicks = {1000L, 1000L, 1001L, 1001L};
        OrbitGenerator waiter = new OrbitGenerator(GeneratorOptions.builder(7)
                .clock(() -> waitTicks[Math.min(waitIndex[0]++, waitTicks.length - 1)])
                .onSequenceExhausted(SequenceExhaustedMode.WAIT)
                .build());
        waiter.restoreState(1000L, 1023);
        long waited = waiter.generate(1);
        assertTrue(waited != 0L);
        assertEquals(1001L, waiter.getLastTimestamp());

        long sample = OrbitId.encode(new OrbitFields(16762354567L, 2, 7, 42));
        assertEquals(16762354567L, OrbitId.getTimestamp(sample));
        assertEquals(2, OrbitId.getType(sample));
        assertEquals(7, OrbitId.getNode(sample));
        assertEquals(42, OrbitId.getSequence(sample));
        assertEquals(16762354567L, OrbitId.getTimestamp(OrbitId.toDecimalString(sample)));
        assertEquals(2, OrbitId.getType(OrbitId.toDecimalString(sample)));
        assertEquals(7, OrbitId.getNode(OrbitId.toDecimalString(sample)));
        assertEquals(42, OrbitId.getSequence(OrbitId.toDecimalString(sample)));
        assertEquals(OrbitId.decode(sample), OrbitId.parse(sample));
        assertTrue(OrbitId.isValid((Object) sample));
        assertFalse(OrbitId.isValid((Object) Boolean.TRUE));
        assertEquals(0L, OrbitId.fromUnixTimeMs(OrbitId.toUnixTimeMs(0L)));

        assertThrows(OrbitError.class, () -> OrbitId.encode(-1L, 1, 1, 0));
        assertThrows(OrbitError.class, () -> OrbitId.encode(1L, 99, 1, 0));
        assertThrows(OrbitError.class, () -> OrbitId.encode(1L, 1, 999, 0));
        assertThrows(OrbitError.class, () -> OrbitId.encode(1L, 1, 1, 9999));

        OrbitError typeZero = assertThrows(OrbitError.class, () -> new OrbitGenerator(7).generate(0));
        assertEquals(OrbitError.INVALID_TYPE, typeZero.getCode());

        OrbitGenerator lost = new OrbitGenerator(GeneratorOptions.builder(1)
                .clock(() -> 5L)
                .confirmOwnership(() -> false)
                .build());
        assertThrows(OrbitError.class, () -> lost.generate(1));
    }



    private static JsonNode fixture(String fileName) throws IOException {
        return JSON.readTree(FIXTURES.resolve(fileName).toFile());
    }
}
